R/log_shiny_input_changes.R

```diff
@@ -37,35 +38,36 @@ log_shiny_input_changes <- function(
     input,
     namespace = NA_character_,
     excluded_inputs = character(),
-    excluded_pattern = "_width$") {
-  session <- shiny::getDefaultReactiveDomain()
+    excluded_pattern = "_width$",
+    session = shiny::getDefaultReactiveDomain()) {
   if (!(shiny::isRunning() || inherits(session, "MockShinySession") || inherits(session, "session_proxy"))) {
     stop("No Shiny app running, it makes no sense to call this function outside of a Shiny app")
   }
-  ns <- ifelse(!is.null(session), session$ns(character(0)), "")
 
-  # utils::assignInMyNamespace and utils::assignInNamespace are needed
-  # so that observer is executed only once, not twice.
-  input_values <- shiny::isolate(shiny::reactiveValuesToList(input))
-  utils::assignInMyNamespace("shiny_input_values", input_values)
+  if (logger::TRACE > logger::as.loglevel(get_val("TEAL.LOG_LEVEL", "teal.log_level", "INFO"))) {
+    # to avoid setting observers when not needed
+    return(invisible(NULL))
+  }
+
+  ns <- ifelse(!is.null(session), session$ns(character(0)), "")
+  reactive_input_list <- shiny::reactive({
+    input_list <- shiny::reactiveValuesToList(input)
+    input_list[!grepl(excluded_pattern, names(input_list))]
+  })
+  shiny_input_values <- shiny::reactiveVal(shiny::isolate(reactive_input_list()))
 
-  shiny::observe({
-    old_input_values <- shiny_input_values
-    new_input_values <- shiny::reactiveValuesToList(input)
+  shiny::observeEvent(reactive_input_list(), {
+    old_input_values <- shiny_input_values()
+    new_input_values <- reactive_input_list()
     names <- unique(c(names(old_input_values), names(new_input_values)))
```

Comment:
`c()` or `intersect()`?
because this is an union operation it might happen that the `old_input_values[name]` might break (*) in case of `name` that exists in `new_input` and missing in `old_input`

(*) in R however, the NULL value will be returned
```
r$> list(a = 1, b = 2, c = 3)['aa']
$<NA>
NULL


r$> list(a = 1, b = 2, c = 3)[['aa']]
NULL
```
so we have to have additional protections against NULL

Reply:
`union` was to register cases when `input[[<name>]]` didn't exist before, to log `input changed from NULL -> <value>`. I'm not a biggest fan of logging input changes at all therefore I'm happy with any logs reduction ;)